src/multicore2.rs

```diff
@@ -0,0 +1,132 @@
+use std::collections::VecDeque;
+use std::sync::Arc;
+
+use parking_lot::{Condvar, Mutex};
+
+use crossbeam::channel::{self, Receiver};
+
+type Task = Box<dyn FnOnce() + Send + 'static>;
+
+#[derive(Default)]
+struct TaskQueue(VecDeque<Task>, VecDeque<Task>);
+
+#[derive(Default)]
+struct State(Mutex<TaskQueue>, Condvar);
+
+#[derive(Clone)]
+pub struct Worker {
+    pub(crate) cpus: usize,
+    shared_state: Arc<State>,
+}
+
+impl Worker {
+    fn spawn_in_pool(shared_state: Arc<State>) {
+        std::thread::spawn(move || {
+            let State(ref mutex, ref cv) = &*shared_state;
+            let mut guard = mutex.lock();
+
+            loop {
+                while let Some(task) = guard.0.pop_front() {
+                    drop(guard);
+                    task();
```

Comment:
What stops me to spawn infinite number of non-cooperative threads in the task?

Reply:
Nothing, there're a lot of ways of halting any threadpool